### ISSUE TITLE
Adopt a websocket from an opened socket file descriptor

### ISSUE THREAD

**Issue by shoedrip** [boostorg/beast#2706]
Hello,

I have a use case where a primary server process handles initial websocket connections, and receive some data from a client then transfers the internal file descriptor socket to another process (along with whatever information from the initial upgrade request). I want to be able to pick that file descriptor from the sub-process and continue exchanging messages with the client transparently.

I can easily create a socket from the native handle, then make a websocket::stream out of it, but I can't do any read/writes until the websocket is in an "open" state, which it can only transition to when calling .accept on it, which is bad because it always sends a 101 Switching Protocols payload when calling that method, even though from the client's perspective the websocket connection was already opened and messages were already exchanged.

Basically I need a way to signal to boost.beast that the websocket::stream is already "opened", that the protocol has already switched, or the ability to do a "dry accept", accepting without sending any data to the client.
Am I missing something?

Thanks.

**Comment by klemens-morgenstern** [comment]
I don't think this is officially supported. 

What you could try (but that's a hack) is open a connection, initialize it, and then switch out the descriptor of the subprocess. Do however note that the websocket negotiates terms of the connection, so you must be absolutely sure they're the same (which is why beast doesn't support this).

This might also be a feature that could be added in the future if there are enough users. Your use case seems quite extra-ordinary.

**Comment by shoedrip** [comment]
I see. I'm not sure about how to do the switcheroo after the websocket has been created though. I know I can call assign on a socket, but after the socket is moved to construct a websocket, I don't know how I can change the underlying fd again. Unless maybe doing some tricks with dup/dup2... I'll give it a go.

I thought beast could handle it as I can easily do what I described in nodejs with the ws package, and it doesn't even attempt to be "low-level".

**Comment by klemens-morgenstern** [comment]
`.next_layer()` will give you the underlying asio object. It might be worth supporting, I'll look into it.

**Comment by ashtum** [comment]
> FWIW, I think that's an interesting use-case, forking a separate process specifically to handle a given _session_, which was accepted on the main process, possibly with some initial back-and-forth (to determine for example whether to keep that connection _in-process_, or _fork-it_).

@ddvienne, could you please elaborate on the use cases for this?
